### ISSUE TITLE
Beast runs over subgrids + merge tool

### ISSUE THREAD

**Issue by drvdputt** [BEAST-Fitting/beast#189]
In light of parallelizing the computation, both for speed as well as memory reasons, it would make sense to do multiple runs in parallel for a set of sub-grids and combine the results at the end. Something similar has already been suggested in #38.  

_Here is an elaborate example for a part of the hypothetical merge:_

To merge some of the statistics, the expectation values for example, I think 1 extra value needs to be stored in the stats file (for each star, to be noted). This value is `sum(exp(lnp_i))`, which is the total, non-normalized weight of the grid, with `lnp_i` the log probabilities. Of course, the `lnp_i` values are typically too large to be able to calculate this value directly, but we can apply the following trick:

In the current code, the weights are first rescaled with an exponent `N_i`:
  `w_i = exp(lnp_i - N_i)`  

After that the sum of the weights `s_i = sum(w_i)` can be taken, and the `w_i` are normalized to 1.
  `w_i /= s_i`  

Therefore, `N_i + s_i = log(sum(exp(lnp_i))) = n_i`, which is the value we're looking for.  The partial expectation values `E_1` and `E_2` for some quantity `E` can then be combined using
  `E_total = (E_1 * exp(n_1 - N) + E_2 * exp(n_2 - N)) / (exp(n_1 - N) + exp(n_2 - N))`  

Where the exponents have again been rescaled with a common term `N`, to make the calculation tractable.

So to merge the results of fits on partial grids, we just need to store this value for each subgrid.

I'm not quite sure what happens when priors are involved though. I can see problems happening when the prior weights depend on the bin widths of one of the parameters.

**Comment by drvdputt** [comment]
I have been doing some more development on this (https://github.com/drvdputt/beast/tree/subgridding) and wrote some code that splits the spectral grid into pieces after the stellar priors have been applied. Then, each of these spectral grid pieces can be turned into extinguished SED grids individually (in a custom run script).

This leaves me with a bunch of sub-sedgrids, for which the noisemodel, trimming and fitting can be done individually, and results in separate lnp, stats, and pdf1d files. I'm now looking for a way to merge the results.

A basic workflow is described in this post: https://github.com/BEAST-Fitting/beast/issues/38#issue-209310882. There is one issue which is not addressed there however: the bin values depend on the minimum and maximum value of the relevant quantity. These can be different for each subgrid (especially the those for the fluxes), and therefore we can not simply sum these 1dpdfs.

Some solutions I have in mind:
- Write out the whole lnp, for each subgrid. We can then recalculate the relevant quantities (such as the pdf1d ranges), by iterating over the different subgrids one by one.
  + expensive (time and disk space), and we basically calculate all the stats twice
  + but will solve most problems.
- Before fitting, look at all the subgrids, and derive the necessary mins/maxes.
  + easier on disk space/IO
  + need to load all the subgrids, for fitting just one subgrid  
- Write the necessary information (mins/maxes) for a subgrid to a file after its construction
  + each process can simply append to the same file, and combine the mins/maxes
  + the contents of this file might have to change as the implementation changes

I think I will be trying the third one, as it seems the most economical.

**Comment by karllark** [comment]
Not quite sure I fully understand this issue.  If the splitting is done for a dimension where the grid is uniform, then putting things back together should not need any other information.  The trimming of grids should not affect this as the trimming is done by only removing models that will give likelihoods that will be zero (or really close).

Can you give an example?

**Comment by drvdputt** [comment]
I originally thought of this issue when I was thinking about what would happen if the grid is split arbitrarily.

But of course, if we split the grid into individual distances for example, then merging the 1dpdfs for the fit parameters themselves should not be a problem.

However, we also calculate and store 1dpdfs for the fluxes in each of the bands (see the below example of a pdf1d file, where the fluxes have 50 bins each). When we split by distance, the range of fluxes, and by consequence the bin centers, will be different for each subgrid.

```
In [23]: pdf1d.info()
Filename: beast_example_phat_pdf1d.gridsub11.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU       6   (2, 2)   float64   
  1  Av            1 ImageHDU         8   (6, 270)   float64   
  2  M_act         1 ImageHDU         8   (49, 270)   float64   
  3  M_ini         1 ImageHDU         8   (50, 270)   float64   
  4  Rv            1 ImageHDU         8   (5, 270)   float64   
  5  Rv_A          1 ImageHDU         8   (9, 270)   float64   
  6  Z             1 ImageHDU         8   (2, 270)   float64   
  7  distance      1 ImageHDU         8   (1, 270)   float64   
  8  f_A           1 ImageHDU         8   (4, 270)   float64   
  9  logA          1 ImageHDU         8   (1, 270)   float64   
 10  logHST_ACS_WFC_F475W_nd    1 ImageHDU         8   (50, 270)   float64   
 11  logHST_ACS_WFC_F475W_wd    1 ImageHDU         8   (50, 270)   float64   
 12  logHST_ACS_WFC_F814W_nd    1 ImageHDU         8   (50, 270)   float64   
 13  logHST_ACS_WFC_F814W_wd    1 ImageHDU         8   (50, 270)   float64   
 14  logHST_WFC3_F110W_nd    1 ImageHDU         8   (50, 270)   float64   
 15  logHST_WFC3_F110W_wd    1 ImageHDU         8   (50, 270)   float64   
 16  logHST_WFC3_F160W_nd    1 ImageHDU         8   (50, 270)   float64   
 17  logHST_WFC3_F160W_wd    1 ImageHDU         8   (50, 270)   float64   
 18  logHST_WFC3_F275W_nd    1 ImageHDU         8   (50, 270)   float64   
 19  logHST_WFC3_F275W_wd    1 ImageHDU         8   (50, 270)   float64   
 20  logHST_WFC3_F336W_nd    1 ImageHDU         8   (50, 270)   float64   
 21  logHST_WFC3_F336W_wd    1 ImageHDU         8   (50, 270)   float64   
 22  logL          1 ImageHDU         8   (50, 270)   float64   
 23  logT          1 ImageHDU         8   (50, 270)   float64   
 24  logg          1 ImageHDU         8   (50, 270)   float64   
 25  mbolmag       1 ImageHDU         8   (50, 270)   float64   
 26  radius        1 ImageHDU         8   (50, 270)   float64   
 27  logHST_WFC3_F275W_wd_bias    1 ImageHDU         8   (50, 270)   float64   
 28  logHST_WFC3_F336W_wd_bias    1 ImageHDU         8   (50, 270)   float64   
 29  logHST_ACS_WFC_F475W_wd_bias    1 ImageHDU         8   (50, 270)   float64   
 30  logHST_ACS_WFC_F814W_wd_bias    1 ImageHDU         8   (50, 270)   float64   
 31  logHST_WFC3_F110W_wd_bias    1 ImageHDU         8   (50, 270)   float64   
 32  logHST_WFC3_F160W_wd_bias    1 ImageHDU         8   (50, 270)   float64   

```

**Comment by karllark** [comment]
Ahh, that now makes sense.  Humm....  Ok.  I see what you are trying to do now.  It is not needed for all the 1d ppdfs, but for just the flux ones, right?  Or is this needed for most given the trimming?

**Comment by drvdputt** [comment]
I think that in some edge cases, the trimming might cut away a value of one of the fit parameters.

For large distances, it might occur that all the models with the largest extinction might be cut out because they are too faint, while for smaller distances, some of these models could still be included.

That would leave us with a smaller range of Av's for large distances, as the grids do not remember their original ranges (the mins and maxes are just recalculated over the whole grid when they are needed).

So for just the fit parameters, I can see it always working without trimming, and going wrong occasionally with trimming.

**Comment by karllark** [comment]
How about determining what the "correct" min/max values are for *all* 1d ppdfs from the full grid and then using these in all the subgrids?  Then the 1d ppdfs would be straightforward to combine in post-processing.  Would need to use the "correct" min/max values in the setup of the 1d ppdfs for every subgrid.  I think this may make it much easier in the long run.  Thoughts?

**Comment by karllark** [comment]
Most excellent.  Sounds like the split and merge of subgrids is getting close to being fully functional.  Yeah!

**Comment by karllark** [comment]
Closing as this should be addressed in the pull request.

**Comment by lea-hagen** [comment]
[Adding comment here, since it seems better than opening a new issue]

**_TL;DR:_** I did a run in which I used the same grid, with and without subgridding, to fit stars in IC1613.  They match!

**_Slightly more detailed description_**

I used this model grid (fairly sparse sampling to make it quick to run):
* log age = [6.0, 10.13, 0.5]
* log metallicities of -2.1, -1.5, and -0.9
* Av = [0.01, 7.0, 0.15]
* Rv = [2.0, 6.0, 1.0]
* fA = [0.0, 1.0, 0.25]

I split the catalogs by source density for both runs.  In the subgridded run, I used 3 subgrids.  The plot below is 5000 randomly chosen stars (out of ~66k).  Each row is a parameter, and the columns are `Best`/`Exp`/`p50` from the stats file. The red dotted line is `x=y`.  If there appear to be any offsets or scatter, it's just from the viewer.

![subgrid_compare](https://user-images.githubusercontent.com/32465297/63286224-889de980-c285-11e9-8ff4-c3fb365dda7d.png)

**Comment by karllark** [comment]
Most excellent!!!!!!!!!
